File: datasets/lua/HumanEval_65_circular_shift.lua
-- Circular shift the digits of the integer x, shift the digits right by shift
-- and return the result as a string.
-- If shift > number of digits, return digits reversed.
-- >>> circular_shift(12, 1)
-- "21"
-- >>> circular_shift(12, 2)
-- "12"
local function circular_shift(x, shift)
  local s = tostring(x)
  local n = #s
  if shift > n then
    return string.reverse(s)
  end
  local s1 = string.sub(s, shift+1, n)
  local s2 = string.sub(s, 1, shift)
  return s1 .. s2
end
-- Unit tests below

lu = require('luaunit')

function test_humaneval()
local candidate = circular_shift
    lu.assertEquals(candidate(100, 2), '001')
    lu.assertEquals(candidate(12, 2), '12')
    lu.assertEquals(candidate(97, 8), '79')
    lu.assertEquals(candidate(12, 1), '21')
    lu.assertEquals(candidate(11, 101), '11')
end

os.exit(lu.LuaUnit.run())

File: datasets/rs/HumanEval_65_circular_shift.rs
/// Circular shift the digits of the integer x, shift the digits right by shift
/// and return the result as a string.
/// If shift > number of digits, return digits reversed.
/// >>> circular_shift(12, 1)
/// "21"
/// >>> circular_shift(12, 2)
/// "12"
fn circular_shift(x: isize, shift: isize) -> String {
    let mut digits = digits(x);
    let len = digits.len();
    if shift > len {
        digits.reverse();
        return digits.into_iter().collect();
    }
    let mut shifted = Vec::with_capacity(len);
    for i in 0..len {
        let j = (i + shift) % len;
        shifted.push(digits[j]);
    }
    shifted.into_iter().collect()

}

fn main() {
    let candidate = circular_shift;
    assert_eq!(candidate(100, 2), String::from("001"));
    assert_eq!(candidate(12, 2), String::from("12"));
    assert_eq!(candidate(97, 8), String::from("79"));
    assert_eq!(candidate(12, 1), String::from("21"));
    assert_eq!(candidate(11, 101), String::from("11"));
}
